crates/sui-framework/packages/bridge/sources/treasury.move

```diff
@@ -2,86 +2,217 @@
-            4
-        } else {
-            abort EUnsupportedTokenType
-        }
+    public struct BridgeTokenMetadata has store, copy, drop {
+        id: u8,
+        decimal_multiplier: u64,
+        notional_value: u64,
+        native_token: bool
     }
 
-    public fun decimal_multiplier<T>(): u64 {
-        let coin_type = type_name::get<T>();
-        if (coin_type == type_name::get<BTC>()) {
-            btc::multiplier()
-        } else if (coin_type == type_name::get<ETH>()) {
-            eth::multiplier()
-        } else if (coin_type == type_name::get<USDC>()) {
-            usdc::multiplier()
-        } else if (coin_type == type_name::get<USDT>()) {
-            usdt::multiplier()
+    public struct ForeignTokenRegistration has store {
+        type_name: TypeName,
+        uc: UpgradeCap,
+        decimal: u8,
+    }
+
+    public struct UpdateTokenPriceEvent has copy, drop {
+        token_id: u8,
+        new_price: u64,
+    }
+
+    public struct NewTokenEvent has copy, drop {
+        token_id: u8,
+        type_name: TypeName,
+        native_token: bool
+    }
+
+    public struct TokenRegistrationEvent has copy, drop {
+        type_name: TypeName,
+        decimal:u8,
+        native_token: bool
+    }
+
+    public(package) fun register_foreign_token<T>(self: &mut BridgeTreasury, tc: TreasuryCap<T>, uc: UpgradeCap, metadata: &CoinMetadata<T>) {
+        // Make sure TreasuryCap has not been minted before.
+        assert!(coin::total_supply(&tc) == 0, ETokenSupplyNonZero);
+        let type_name = type_name::get<T>();
+        let coin_address = address::from_ascii_bytes(ascii::as_bytes(&type_name::get_address(&type_name)));
+        // Make sure upgrade cap is for the Coin package
+        assert!(object::id_to_address(&package::upgrade_package(&uc)) == coin_address, EInvalidUpgradeCap);
+        let registration = ForeignTokenRegistration {
+            type_name,
+            uc,
+            decimal: coin::get_decimals(metadata),
+        };
+        bag::add(&mut self.waiting_room, type_name::into_string(type_name), registration);
+        object_bag::add(&mut self.treasuries, type_name, tc);
+
+        emit(TokenRegistrationEvent{
+            type_name,
+            decimal: coin::get_decimals(metadata),
+            native_token: false
+        });
+    }
+
+    public fun token_id<T>(self: &BridgeTreasury): u8 {
+        let metadata = self.get_token_metadata<T>();
+        metadata.id
+    }
+
+    public fun decimal_multiplier<T>(self: &BridgeTreasury): u64 {
+        let metadata = self.get_token_metadata<T>();
+        metadata.decimal_multiplier
+    }
+
+    public fun notional_value<T>(self: &BridgeTreasury): u64 {
+        let metadata = self.get_token_metadata<T>();
+        metadata.notional_value
+    }
+
+    public(package) fun approve_new_token(self: &mut BridgeTreasury, token_name: String, token_id:u8, native_token: bool, notional_value: u64) {
```

Comment:
I think we should name this `add_new_token` because this function does not really "approve" anything